twine/commands/check.py

```diff
@@ -68,47 +69,74 @@ def __str__(self):
         return self.output.getvalue()
 
 
+def _get_description_content_type(metadata, output_stream):
+    description_content_type = metadata["description_content_type"]
+    if description_content_type is None:
+        output_stream.write(
+            'warning: `long_description_content_type` missing.  '
+            'defaulting to `text/x-rst`.\n'
+        )
+        description_content_type = 'text/x-rst'
+
+    if description_content_type not in _VALID_CONTENT_TYPES:
+        output_stream.write(
+            'warning; `long_description_content_type` invalid.\n'
+            'It must be one of the following types: [{}].\n'
+            .format(", ".join(_VALID_CONTENT_TYPES))
+        )
+    return description_content_type
+
+
+def _get_description(metadata, output_stream):
+    description = metadata["description"]
+    if description in {None, 'UNKNOWN\n\n\n'}:
+        output_stream.write('warning: `long_description` missing.\n')
+        return None
+    return description
+
+
+def _description_will_not_render(description_content_type, description):
+    warning_stream = _WarningStream()
+    content_type, params = cgi.parse_header(description_content_type)
+    renderer = _RENDERERS.get(content_type, _RENDERERS[None])
+    return (description is not None
+            and renderer
+            and renderer.render(description, stream=warning_stream, **params)
+            is None)
+
+
+def check_package(package, output_stream):
+    warning_stream = _WarningStream()
+    metadata = package.metadata_dictionary()
+    description_content_type = _get_description_content_type(
+        metadata,
+        output_stream
+    )
+    description = _get_description(metadata, output_stream)
+    failure = _description_will_not_render(
```

Comment:
Right now a render failure is the only thing that triggers a a failure. I personally think the description type being provided but invalid should be a failure as well but I wanted to wait for feedback before I did that. See https://github.com/pypa/twine/pull/473#discussion_r300879410

Reply:
I tend to agree, but maybe there are some other considerations that we don't know about.

With that in mind, I think it will be easier to adjust the failure modes if the utility functions above were inlined here, as they were in `check()`. It would also be less of a structural change.

Related to that, the `_get_*` functions feel like they're doing two things: retrieving a value, and checking its validity. If they don't get inlined, I think it'd be more flexible if they were just validators (possibly with the warning output), similar to `_description_will_not_render`.

What do you think?

Reply:
Im not sure. They warn you if something is potentially strange but its basically just a log line. The `check_package` function does not really depend on any of that  that other than just getting the value. That's why I see them as getters. 

If the validation was actually relevant to the check_package function I could see either separating it out or inlining that validation to `check_package`

actually saying that makes me wanna go ahead and make the type check a proper validation. Since unlike the other two warnings. It actually does matter.

Reply:
I went ahead and inlined the getters. What pushed me over the edge was realizing the order which output stream gets written to does matter. Given this I did not want to be passing around output_stream and having functions return both their values and any warnings they wanted to emit felt like a bridge too far.